src/drivers/intel/hda/hda-dma.c

```diff
@@ -587,15 +583,15 @@ static int hda_dma_pause(struct dma_chan_data *channel)
 {
 	uint32_t flags;
 
-	irq_local_disable(flags);
-
 	tr_dbg(&hdma_tr, "hda-dmac: %d channel %d -> pause",
 	       channel->dma->plat_data.id, channel->index);
 
+	irq_local_disable(flags);
 	if (channel->status != COMP_STATE_ACTIVE)
 		goto out;
 
 	/* stop the channel */
+	irq_local_enable(flags);
```

Comment:
this doesn't look right - `irq_local_enable()` is called twice here. I think here we either have to protect the `.status` check above and `hda_dma_stop()` call below together or we don't have to lock at all. Locking only the check makes no sense because if an interrupt can change the `.status`, it anyway can happen right after re-enabling interrupts. So if that is possible, we have to lock the interrupts for both. `hda_dma_stop()` disables interrupts internally too, which doesn't hurt, but is redundant in this case. So we could make an "unlocked" version of that function, but that would be a minor optimisation.